packages/language-support/src/helpers.ts

```diff
@@ -37,11 +56,70 @@ export function findStopNode(root: StatementsContext) {
   return current;
 }
 
+export function findNearestNodeStartingBeforeOrAt(
+  current: ParseTree,
+  position: number,
+): EnrichedParseTree | undefined {
+  let result: EnrichedParseTree | undefined = undefined;
+
+  if (
+    current instanceof TerminalNode &&
+    current.symbol.type !== CypherParser.EOF
+  ) {
+    const symbol = current.symbol;
+
+    if (symbol.start <= position) {
+      result = current;
+    }
```

Comment:
If I read this function correctly, it recursively walks the three left child first, and stops assigning the "result" variable when the current leaf being processed is no longer having lower result than the position we looked for. 

So in total right now we're walking the tree from the top down to the closest child to the cursor, then we walk it back up again to find a `FunctionInvocationContext` an `CallClauseContext` or an `ExpressionContext` with a `(` error node. It seems to me the classic "antlr" way to do this is with a parse visitor for those contexts, did you consider that and if so why do you prefer the "manual" way?